### PR TITLE
V2.1.5

### PR DESCRIPTION
**Pull request by Marios85**
This version includes the bug fix for issue #108 - Unhandled exception when SSH key is missing (or any other git exception when downloading the code from the repo).

Please note that MyPy errors had to be ignored on a few lines because it has a bug that means it can't calculate short circuiting of IF statements. I upgraded the MyPy version to the latest but this did not resolve the issue. As a result Flake8 has to be ignored on those lines as well because there is no option but to have the declaration on the exact line (you can ignore at the top of the file but this introduced a bug where a file could not find the run_workflow method, yet another bug with MyPy).

### CODE REVIEW

**Review by AlexIoannides — APPROVED** [review]
Nicely done (as always).

Small query on the need to ignore MyPy (more of a curiosity).

File: src/bodywork/workflow_execution.py [hunk 1]
```diff
@@ -153,18 +154,22 @@ def run_workflow(
...
+                    BodyworkNamespaceError,
+                    BodyworkDockerImageError,
+                    BodyworkGitError,
+                    BodyworkConfigError,
+                ]
+                and config.project.run_on_failure   # type: ignore  # noqa
+            ):
                 _run_failure_stage(
-                    config, e, namespace, repo_url, repo_branch, docker_image
+                    config, e, namespace, repo_url, repo_branch, docker_image   # type: ignore  # noqa
```
[the remark this answers is not in the document]
Comment (Marios85) [reply src/bodywork/workflow_execution.py:168]: The same error as before, saying that config might not have been assigned even though it is impossible to get there with an uninitialised config. 

Black ignores the comments with regards to line length, doesn't do anything about it. I wonder if we really should be using MyPy, its not even on version one so it must mean its Alpha/Beta and it shows, there's multiple bugs.

[the remark this answers is not in the document]
Comment (AlexIoannides) [reply src/bodywork/workflow_execution.py:168]: MyPy is not some small flakey project and has been running for years. The version number is irrelevant - it's just the versioning scheme they use.

If I remove this line from `workflow_execution.py`:

image

Then all the checks pass with the `# type: ignore  # noqa` tags **removed**.

I know it's not always perfect form a user-experience perspective, but 99% of the time it's right and leads to more robust code. The only alternative is not to use type-hints...

[the remark this answers is not in the document]
Comment (Marios85) [reply src/bodywork/workflow_execution.py:168]: If you remove that then you'll cause an exception to be raised in the finally block as there is a check to see if the config is present for the usage stats. That's the reason its there. 

I've not had a good experience of it so far, in addition to this, adding the ignore line to the top of the file to ignore the whole file from file checking causes the run_workflow method to not be found in an import statement in cli.py. That seems pretty dodgy to me.

[the remark this answers is not in the document]
Comment (AlexIoannides) [reply src/bodywork/workflow_execution.py:168]: You're right.

I've just submitted an alternative in #110, which I think solves the issues and leaves cleaner code.

**Pull request merged**